File: joshua/BinaryTreeInorderTraversal.java
package com.lintcode;

import java.util.ArrayList;
import java.util.Stack;

/**
 * @author : Joshua
 * Date:     2/1/16
 * @see <a href="http://www.lintcode.com/zh-cn/problem/binary-tree-inorder-traversal/">二叉树的中序遍历</a>
 */
public class BinaryTreeInorderTraversal {
    public static class TreeNode {
        public int val;
        public TreeNode left, right;
        public TreeNode(int val) {
            this.val = val;
            this.left = this.right = null;
        }
    }

    public ArrayList<Integer> inorderTraversal(TreeNode root) {
        // write your code here
        ArrayList<Integer> resultList = new ArrayList<Integer>();
        if (root != null) {
            fillList(resultList, root);
        }
        return resultList;
    }

    private void fillList(ArrayList<Integer> treeNodeList, TreeNode node) {
        Stack<TreeNode> nodeStack = new Stack<TreeNode>();
        while (node != null || !nodeStack.isEmpty()) {
            while (node != null) {
                TreeNode t = node;
                node = node.left;
                nodeStack.push(t);
            }
            node = nodeStack.pop();
            treeNodeList.add(node.val);
            node = node.right;
        }
    }
}

File: lulu/binary_tree_inorder_traversal.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
    """
    @param root: The root of binary tree.
    @return: Inorder in ArrayList which contains node values.
    """
    def inorderTraversal(self, root):
        # write your code here
        result = []
        self.in_order_traversal(root, result)
        return result

    def in_order_traversal(self, root, result):
        if root is None:
            return
        self.in_order_traversal(root.left, result)
        result.append(root.val)
        self.in_order_traversal(root.right, result)
